### PR TITLE
[Tier2] Redploy non-collocated OSDs

### PR DESCRIPTION
**Pull request by adityaramteke**
- Create non-collocated OSDs
- Zap multiple OSDs
- Verify OSD creation is successful
-  Validate LVM for each OSD
- OSD db should not be created on a new device.

Logs: http://magna002.ceph.redhat.com/cephci-jenkins/aramteke/automation-83575600/cephci-run-GI7DO4/

### CODE REVIEW

File: ceph/ceph.py [hunk 1]
```diff
@@ -47,6 +47,7 @@ def __init__(self, name, node_list=None):
         self.custom_config = None
```
Comment (vamahaja) [inline ceph/ceph.py:47]: this change should go here - https://github.com/red-hat-storage/cephci/pull/3157

File: cli/utilities/configure.py [hunk 2]
```diff
@@ -329,11 +329,9 @@ def setup_installer_node(
     # Check for ansible preflight
     if not ansible_preflight:
         # Install cephadm package
-        install_cephadm(
-            installer, ceph_version, platform, tools_repo, build_type, ibm_build
-        )
+        install_cephadm(installer, build_type)
 
-        return
```
Comment (vamahaja) [inline cli/utilities/configure.py:336]: Why this change ?

File: cli/utilities/configure.py [hunk 3]
```diff
@@ -329,11 +329,9 @@ def setup_installer_node(
     # Check for ansible preflight
     if not ansible_preflight:
         # Install cephadm package
-        install_cephadm(
-            installer, ceph_version, platform, tools_repo, build_type, ibm_build
```
Comment (vamahaja) [inline cli/utilities/configure.py:333]: is method signature changed ?

File: run.py [hunk 4]
```diff
@@ -237,6 +237,7 @@ def create_nodes(
                     private_ip=private_ip,
```
Comment (vamahaja) [inline run.py:237]: this change should go here - https://github.com/red-hat-storage/cephci/pull/3157

File: suites/quincy/ceph_volume/tier-2-ceph-volume.yaml [hunk 5]
```diff
@@ -87,3 +87,4 @@ tests:
       module: test_ceph_volume_lvm_tags_after_osd_rm.py
       config:
         operation: lvm-zap
```
Comment (vamahaja) [inline suites/quincy/ceph_volume/tier-2-ceph-volume.yaml:89]: Why this change ?

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: suites/quincy/ceph_volume/tier-2-ceph-volume.yaml (hunk 5 above)
Comment (adityaramteke) [reply suites/quincy/ceph_volume/tier-2-ceph-volume.yaml]: Reverted the changes

File: tests/ceph_volume/test_ceph_volume_rm.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+from ceph.waiter import WaitUntil
+from cephci.utils.configs import get_cloud_credentials, get_configs
+from cli.cephadm.cephadm import CephAdm
+from cli.cloudproviders import CloudProvider
+from cli.cluster.node import Node
+from cli.cluster.volume import Volume
+from cli.utilities.utils import get_lvm_on_osd_container, get_running_containers
+from utility.log import Log
+
+log = Log(__name__)
```
Comment (vamahaja) [inline tests/ceph_volume/test_ceph_volume_rm.py:15]: Unused variable

File: tests/ceph_volume/test_ceph_volume_rm.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+            Volume(volname, cloud).create(
+                size=device_spec["db_devices"]["size"].replace("G", "")
+            )
+            Node(osd_node.ceph_nodename, cloud).attach_volume(volname)
+
+    # Define CephAdm class object with mount file
+    cephadm = CephAdm(nodes=installer, mount=file.name)
+
+    # Create OSDs with yaml file
+    cmd = f"ceph orch apply -i {file.name}"
```
Comment (vamahaja) [inline tests/ceph_volume/test_ceph_volume_rm.py:75]: Use - https://github.com/red-hat-storage/cephci/blob/master/cli/ceph/orch/orch.py#L53

File: tests/ceph_volume/test_ceph_volume_rm.py [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    config = configs.get("config")
+
+    # Create temp yaml file
+    file = tempfile.NamedTemporaryFile(suffix=".yaml")
+
+    # Get the installer node
+    installer = ceph_cluster.get_nodes(role="installer")[0]
+
+    # Write data to file
+    with installer.remote_file(sudo=True, file_name=file.name, file_mode="w") as f:
```
Comment (vamahaja) [inline tests/ceph_volume/test_ceph_volume_rm.py:36]: instead use method - https://github.com/red-hat-storage/cephci/blob/master/cli/ops/cephadm.py#L44

File: tests/ceph_volume/test_ceph_volume_rm.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    # Get device spec
+    device_spec = configs["spec"]["spec"]
+
+    # Get cloud type
+    cloud_type = configs.get("cloud-type")
+
+    # Get OSD nodes
+    osd_nodes = ceph_cluster.get_nodes(role="osd")
+
+    get_configs(config)
```
Comment (vamahaja) [inline tests/ceph_volume/test_ceph_volume_rm.py:48]: Move this to start of the test case

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: tests/ceph_volume/test_ceph_volume_rm.py (hunk 6 above)
Comment (adityaramteke) [reply tests/ceph_volume/test_ceph_volume_rm.py:13]: I have now been using it

File: tests/ceph_volume/test_ceph_volume_rm.py (hunk 7 above)
Comment (adityaramteke) [reply tests/ceph_volume/test_ceph_volume_rm.py]: Updated

File: tests/ceph_volume/test_ceph_volume_rm.py (hunk 8 above)
Comment (adityaramteke) [reply tests/ceph_volume/test_ceph_volume_rm.py]: Updated

File: tests/ceph_volume/test_ceph_volume_rm.py (hunk 9 above)
Comment (adityaramteke) [reply tests/ceph_volume/test_ceph_volume_rm.py]: Updated

File: tests/ceph_volume/test_ceph_volume_redeploy.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+from cli.cluster.node import Node
+from cli.cluster.volume import Volume
+from cli.ops.cephadm import generate_bootstrap_config
+from cli.utilities.utils import get_lvm_on_osd_container, get_running_containers
+from utility.log import Log
+
+log = Log(__name__)
+
+
+class RemoveCephVolumeError(Exception):
```
Comment (vamahaja) [inline tests/ceph_volume/test_ceph_volume_redeploy.py:16]: Use common exception - https://github.com/red-hat-storage/cephci/blob/master/cli/exceptions.py#L32

File: tests/ceph_volume/test_ceph_volume_redeploy.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+    # Get the lvm list afer OSD zap
+    lvm_list = get_lvm_on_osd_container(container_ids[0], osd_nodes[0])
+
+    # Get OSD and device after zap
+    osd_db_after = _get_osd_db_id(osd_ids=osd_ids, lvm_list=lvm_list)
+
+    # Validate if db device has changed
+    if osd_db_before != osd_db_after:
+        raise RemoveCephVolumeError("Faild to re-deploy non-collocated OSD")
+    else:
```
Comment (vamahaja) [inline tests/ceph_volume/test_ceph_volume_redeploy.py:124]: This else statement is not required here

File: tests/ceph_volume/test_ceph_volume_redeploy.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+    # Get the installer and OSD nodes
+    installer = ceph_cluster.get_nodes(role="installer")[0]
+    osd_nodes = ceph_cluster.get_nodes(role="osd")
+
+    # Get device spec
+    device_spec = config.get("spec", {}).get("spec", {})[0].get("spec", {})
+
+    # Get cloud creadentials
+    cloud_type = config.get("cloud-type")
```
Comment (vamahaja) [inline tests/ceph_volume/test_ceph_volume_redeploy.py:66]: These variables can be moved to method `_create_volume_in_bulk`

File: tests/ceph_volume/test_ceph_volume_redeploy.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+from utility.log import Log
+
+log = Log(__name__)
+
+
+class RemoveCephVolumeError(Exception):
+    pass
+
+
+def _create_volume_in_bulk(nodes, cloud, size, count):
```
Comment (vamahaja) [inline tests/ceph_volume/test_ceph_volume_redeploy.py:20]: Better to use name `_add_node_volumes`

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: tests/ceph_volume/test_ceph_volume_redeploy.py (hunk 11 above)
Comment (adityaramteke) [reply tests/ceph_volume/test_ceph_volume_redeploy.py:124]: Okay, I'll remove it.

File: tests/ceph_volume/test_ceph_volume_redeploy.py (hunk 12 above)
Comment (adityaramteke) [reply tests/ceph_volume/test_ceph_volume_redeploy.py:66]: Done, updated.

File: tests/ceph_volume/test_ceph_volume_redeploy.py (hunk 13 above)
Comment (adityaramteke) [reply tests/ceph_volume/test_ceph_volume_redeploy.py:20]: Done, updated.

File: conf/reef/ceph_volume/tier-2_cephvolume_5node_one_volume.yaml [hunk 14]
```diff
@@ -0,0 +1,49 @@
+globals:
```
Comment (vamahaja) [inline conf/reef/ceph_volume/tier-2_cephvolume_5node_one_volume.yaml:1]: Change config name to - `1admin-1clinet-5node-single-volume.yaml`

**Review by vamahaja — CHANGES_REQUESTED** [review]

File: suites/reef/ceph_volume/tier-2-ceph-volume-redeploy.yaml [hunk 15]
```diff
@@ -0,0 +1,86 @@
+# Name: Cluster deployment with Non-collocated OSD
```
[the remark this answers is not in the document]
Comment (adityaramteke) [reply suites/reef/ceph_volume/tier-2-ceph-volume-redeploy.yaml]: Done, updated.

File: conf/reef/ceph_volume/tier-2_cephvolume_5node_one_volume.yaml (hunk 14 above)
Comment (adityaramteke) [reply conf/reef/ceph_volume/tier-2_cephvolume_5node_one_volume.yaml]: Done, updated.

**Review by vamahaja — CHANGES_REQUESTED** [review]
LGTM. Please share result logs

File: tests/ceph_volume/test_ceph_volume_redeploy.py [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+    nodes = nodes if type(nodes) in (list, tuple) else [nodes]
+    for node in nodes:
+        for i in range(count):
+            volname = f"{node.ceph_nodename}-vol-{size}-{i}"
+            Volume(volname, cloud).create(size=size)
+            Node(node.ceph_nodename, cloud).attach_volume(volname)
+    return True
+
+
+def _create_yaml_file(node, config):
```
Comment (vamahaja) [inline tests/ceph_volume/test_ceph_volume_redeploy.py:39]: instead use - https://github.com/red-hat-storage/cephci/blob/master/cli/ops/cephadm.py#L44

**Review by vamahaja — CHANGES_REQUESTED** [review]

**Review by pranavprakash20 — APPROVED** [review]
Lgtm

**Review by tintumathew10 — APPROVED** [review]
lgtm
